### PR TITLE
Buffer timeout interfaces

### PR DESCRIPTION
**Pull request by mroda88**
Details of this PR in #12

### CODE REVIEW

File: include/app-framework-base/Buffers/Buffer.hh [hunk 1]
```diff
@@ -26,43 +27,44 @@ public:
...
 
 public:
-  explicit BufferInput(size_t pt = 10) : fPushTimeout_ms(pt) {}
+  template <typename U> void push(U &&t) {
+    push_wait_for(std::forward<U>(t), DurationType(0));
+  }
 
-  virtual int push(const T &t) { return push(std::move(t)); }
-  virtual int push(T &&) = 0;
+  virtual void push_wait_for(ValueType &&, const DurationType &) = 0;
```
Comment (alessandrothea) [inline include/app-framework-base/Buffers/Buffer.hh:38]: Is the `_wait_for` necessary? I understand the desire of being uber explicit, but  a signature like

`virtual void push(ValueType &&, const DurationType &timeout) = 0;`

is equally clear and compact.

File: include/app-framework-base/Buffers/Buffer.hh [hunk 2]
```diff
@@ -26,43 +27,44 @@ public:
...
-  BufferOutput(const BufferOutput &) = default;
-  BufferOutput &operator=(const BufferOutput &) = default;
-  BufferOutput(BufferOutput &&) = default;
-  BufferOutput &operator=(BufferOutput &&) = default;
+  // in your derived class declaration (line above assumes you're
+  // using the same template parameter labels)
 
-private:
-  size_t fPopTimeout_ms;
+  virtual ValueType pop_wait_for(const DurationType &) = 0;
```
Comment (alessandrothea) [inline include/app-framework-base/Buffers/Buffer.hh:67]: Same comment as above. Preferred signature
`virtual ValueType pop(const DurationType &timeout)`

**Comment by alessandrothea** [comment]
On top of two inline comments, I would also add that the I/O interfaces should expose `full` and `empty` flags.

**Comment by alessandrothea** [comment]
@mroda88 Please put a reference to relevant tickets when opening a PR

**Comment by jcfreeman2** [comment]
-I've removed the "_wait_for" suffixes to the pushes and pops in the buffer interfaces. 

-I'm a little unclear on the comment "I would also add that the I/O interfaces should expose full and empty flags." BufferInput and BufferOutput both inherit the empty() and full() functions declared in their parent class, BufferI. Or am I misunderstanding something?

**Pull request merged**